go/pvl/interp_test.go

```diff
@@ -1,4 +1,647 @@
-// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// Copyright 2016 Keybase, Inc. All rights reserved. Use of
 // this source code is governed by the included BSD license.
 
 package pvl
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/PuerkitoBio/goquery"
+	libkb "github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+	jsonw "github.com/keybase/go-jsonw"
+)
+
+type interpUnitTest struct {
+	name      string
+	proofinfo ProofInfo
+	prepvl    map[keybase1.ProofType]string
+	service   keybase1.ProofType
+
+	// What api call to expect
+	restype     libkb.XAPIResType
+	resjson     string
+	reshtml     string
+	restext     string
+	urloverride string
+	// (Optional) Expect a different url to be fetched than proofinfo.APIURL
+
+	// Whether the proof should validate
+	shouldwork bool
+	// (Optional) error status must equal. Must be specified for INVALID_PVL
+	errstatus keybase1.ProofStatus
+	// (Optional) error string must equal
+	errstr string
+}
+
+var interpUnitTests = []interpUnitTest{
+	// # Basic tests
+	{
+		name:      "basichtml",
+		proofinfo: info1,
+		prepvl: map[keybase1.ProofType]string{
+			keybase1.ProofType_TWITTER: `[
+{"assert_regex_match": "^https://rooter\\.example\\.com/proofs/%{username_service}/[\\d]+\\.htjsxt$"},
+{"fetch": "html"},
+{"selector_css": [".twit", 0]},
+{"whitespace_normalize": true},
+{"assert_regex_match": "^.*goodproof.*$"}
+]`},
+		service:    keybase1.ProofType_TWITTER,
+		restype:    libkb.XAPIResHTML,
+		reshtml:    html1,
+		shouldwork: true,
+	},
+
+	// # Tests for individual valid instructions.
+	// Test match and fail of each instruction.
+
+	// ## AssertRegexMatch,
+	{
+		name:      "AssertRegexMatch-url-ok",
+		proofinfo: info1,
+		prepvl: map[keybase1.ProofType]string{
+			keybase1.ProofType_COINBASE: `[
+{"assert_regex_match": "^https://rooter\\.example\\.com/proofs/%{username_service}/[\\d]+\\.htjsxt$"},
+{"fetch": "string"}
+]`},
+		service:    keybase1.ProofType_COINBASE,
+		restype:    libkb.XAPIResText,
+		restext:    "1234",
+		shouldwork: true,
+	}, {
+		name:      "AssertRegexMatch-url-fail",
+		proofinfo: info1,
+		prepvl: map[keybase1.ProofType]string{
+			keybase1.ProofType_HACKERNEWS: `[
+{"assert_regex_match": "^https://rooter\\.example\\.com/proofs/%{username_service}/+\\.htjsxt$"},
+{"fetch": "string"}
+]`},
+		service:    keybase1.ProofType_HACKERNEWS,
+		restype:    libkb.XAPIResText,
+		restext:    "1234",
+		shouldwork: false,
+	}, {
+		name:      "AssertRegexMatch-content-ok",
+		proofinfo: info1,
+		prepvl: map[keybase1.ProofType]string{
+			keybase1.ProofType_REDDIT: `[
```

Comment:
These are all fake anyway. I tried to vary which ProofType is used just to encourage bugs in any future service-specific code to show up.